torch/distributed/_shard/checkpoint/filesystem.py

```diff
@@ -0,0 +1,127 @@
+import io
+import os
+import operator
+import pickle
+from typing import List, Optional, cast
+
+import torch
+from torch import Tensor
+import torch.distributed as dist
+from torch.futures import Future
+
+from .metadata import (
+    BytesReadRequest,
+    BytesWriteRequest,
+    Metadata,
+    TensorReadRequest,
+    TensorWriteRequest,
+)
+from .storage import StorageReader, StorageWriter
+
+class FileSystemWriter(StorageWriter):
+    def __init__(self, path: str) -> None:
+        super().__init__()
+        self.path = path
+        os.makedirs(self.path, exist_ok=True)
+
+    def write_bytes(self, requests: List[BytesWriteRequest]) -> Future[None]:
+        for req in requests:
+            with open(os.path.join(self.path, req.storage_key), "wb") as storage:
```

Comment:
Especially for the case of handling case-insensitive file systems, maybe it would be better to not use the key at all in the path name.  It seems like it would perhaps be better to just pick arbitrary path names, and have a mapping of storage key to path in the metadata.

This would also make it easier to only save 1 copy of the data if `save_state_dict()` is called with multiple tensors that all point to the same storage.

Reply:
Great point, the handle/file name hashing is essential for production system. I
IUC, we are all talking about doing the name hash inside the StorageWriter/Reader but not above it, right? 
And I assume the comment here is more about the advanced/production case and not about this basic OSS file system.

Reply:
> On your last point, about having multiple tensors pointing to the same storage, how common is to run into models where multiple tensors cover overlapping parts of their underlying storage?

For checkpointing case, I tend to consider it a bug in state_dict. Shall we create a utility function that checks overlapping memory, in state_dict_saver/state_dict_loader, and prevent it from reaching the storagerwriter/reader

Reply:
See T87045442 for one example where we have models with shared storage in the state dict.

I also think we should build a clean API here that makes no assumptions about how it is being used.  I think we should provide a generic key-value store, that doesn't care what the keys are, or how parameter FQNs are constructed.  I think we should make sure the API is completely generic so we don't have to debug problems later on where something got changed higher up and ended up breaking undocumented assumptions in the checkpoint storage code.